lib/metadata/acl.js

```diff
@@ -14,10 +14,21 @@ const acl = {
 
     addObjectACL(bucket, objectKey, objectMD, addACLParams, params, log, cb) {
         log.trace('updating object acl in metadata');
-        // eslint-disable-next-line no-param-reassign
-        objectMD.acl = addACLParams;
-        metadata.putObjectMD(bucket.getName(), objectKey, objectMD, params, log,
-            cb);
+        function _didNotChange(grant) {
+            return objectMD.acl[grant].length === addACLParams[grant].length
+                && objectMD.acl[grant].every(value => addACLParams[grant].includes(value));
```

Comment:
not sure this condition is right : if the new ACL includes the old one, it does not mean it did not change, just that nothing was removed...

if the function did only "adding" ACL (and we were to merge the existing ACL with the new one), I would expect the condition to be the other way round : no change if the new ACL are "included" in the existing ones already...

That said, I am not sure what is expected exacty of this API : so i don't know which behavior should be expected.

Reply:
We have five cases here for each grant:
- new ACL adds new grantees => line 18 would catch that (we check the length of both arrays)
- new ACL removes new grantees => line 18 would also catch that
- new ACL does not change anything => both tests will pass
- new ACL replaces grantees (both addition and deletion is made) => we may have the same number of grantees in some cases but the values would be different and so this will be caught in line 19

Reply:
if both grants (old and new one) contain the same number of elements and if all elements in one of them are included in the second, that means both grants contain the same elements, so if we add new grantees this will be caught in line 18

Reply:
indeed, you are right.

that said, if we need to "think" if the condition is right, then maybe there is something to improve : make implementation easier to read and/or at least a comment...